pkg/controllers/trace_controller.go

```diff
@@ -0,0 +1,247 @@
+				if err != nil {
+					// Print error message but don't try again later
+					log.Errorf("Failed to delete tracer BPF map: %s", err)
+				}
+			}
+
+			// Remove our finalizer
+			controllerutil.RemoveFinalizer(trace, GADGET_FINALIZER)
+			if err := r.Update(ctx, trace); err != nil {
+				log.Errorf("Failed to remove finalizer: %s", err)
+				return ctrl.Result{}, err
+			}
+		}
+		// Stop reconciliation as the Trace is being deleted
+		log.Infof("Let trace %s be deleted", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	// The Trace is not being deleted, so register our finalizer
+	beforeFinalizer := trace.DeepCopy()
+	controllerutil.AddFinalizer(trace, GADGET_FINALIZER)
+	if err := r.Patch(ctx, trace, client.MergeFrom(beforeFinalizer)); err != nil {
+		log.Errorf("Failed to add finalizer: %s", err)
+		return ctrl.Result{}, err
+	}
+
+	// Register tracer
+	if r.TracerManager != nil {
+		_, err = r.TracerManager.AddTracer(ctx,
+			&pb.AddTracerRequest{
+				Id:       gadgets.TraceNameFromNamespacedName(req.NamespacedName),
+				Selector: genSelector(trace.Spec.Filter),
+			})
+		if err != nil && !errors.Is(err, os.ErrExist) {
+			log.Errorf("Failed to add tracer BPF map: %s", err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// For now, only support control via the GADGET_OPERATION
+	if trace.Spec.RunMode != "Manual" {
+		log.Infof("Unsupported RunMode: %q", trace.Spec.RunMode)
+		return ctrl.Result{}, nil
+	}
+
+	// For now, only support output in the TraceStatus field
+	if trace.Spec.OutputMode != "Status" {
+		log.Infof("Unsupported OutputMode: %q", trace.Spec.OutputMode)
+		return ctrl.Result{}, nil
+	}
+
+	// Lookup annotations
+	if trace.ObjectMeta.Annotations == nil {
+		log.Info("No annotations. Nothing to do.")
+		return ctrl.Result{}, nil
+	}
+	var op string
+	if op, ok = trace.ObjectMeta.Annotations[GADGET_OPERATION]; !ok {
+		log.Info("No operation annotation. Nothing to do.")
+		return ctrl.Result{}, nil
+	}
+
+	params := make(map[string]string)
+	for k, v := range trace.ObjectMeta.Annotations {
+		if !strings.HasPrefix(k, GADGET_OPERATION+"-") {
+			continue
+		}
+		params[strings.TrimPrefix(k, GADGET_OPERATION+"-")] = v
+	}
+
+	log.Infof("Gadget %s operation %q on %s", trace.Spec.Gadget, op, req.NamespacedName)
+
+	// Remove annotations first to avoid another execution in the next
+	// reconciliation loop.
+	withAnnotation := trace.DeepCopy()
+	annotations := trace.GetAnnotations()
+	delete(annotations, GADGET_OPERATION)
+	for k, _ := range params {
+		delete(annotations, GADGET_OPERATION+"-"+k)
+	}
+	trace.SetAnnotations(annotations)
+	err = r.Patch(ctx, trace, client.MergeFrom(withAnnotation))
+	if err != nil {
+		log.Errorf("Failed to update trace: %s", err)
+		return ctrl.Result{}, err
+	}
+
+	// Call gadget.Operation()
+	traceBeforeOperation := trace.DeepCopy()
+	patch := client.MergeFrom(traceBeforeOperation)
+	err = gadget.Operation(trace, op, params)
```

Comment:
Maybe I should just remove the return value. I was thinking of allowing the controller to try again but I don't know if that really work because we would have removed the annotation by then... and it's currently not used by any gadget. It could be added back if there is a need, but I don't see a need right now.

Reply:
Updated without the error value; so now the gadget cannot ask the controller to requeue.